### ISSUE TITLE
FeatureFlagsLive.setProvider rollback restores providerRef but not SDK-client routing, so a failed hot-swap can leave evaluations hitting the failed provider

### ISSUE THREAD

**Issue by EtaCassiopeia** [EtaCassiopeia/zio-openfeature#282]
## Problem

In `core/src/main/scala/zio/openfeature/FeatureFlagsLive.scala`, `setProvider`'s failure path (`tapError`) rolls back the internal `providerRef` (and `providerNameRef`) to the previous provider and sets status `Error`. However, evaluation routes through `evaluateFromClient`, which uses the OpenFeature Java SDK `Client` — the SDK routes to whatever provider `setProviderAndWait` last registered, NOT to our `providerRef`. When a swap's `setProviderAndWait(newProvider)` fails (e.g., the new provider's `initialize()` throws), the SDK may leave the failed new provider registered while our `providerRef` points back at the old one — the two disagree, and evaluations continue returning values from the failed provider.

## Evidence

In `fromAcquireAsync`'s AC2b test (PR #281), a forced swap failure left evaluations returning the failed provider's value rather than the fallback's. This confirms the Java SDK client still routed to the failed provider even though `providerRef` had rolled back. (PR #281 worked around the *status* half by force-restoring an evaluable status so evals don't hard-fail — but the routing mismatch itself is pre-existing and unaddressed.)

## Provenance

Found by an agent while implementing #241 / PR #281 (non-blocking provider init). Out of scope for that work, filed as a follow-up.

---

## Verdict: worth fixing

This is a genuine correctness bug, not a cosmetic mismatch: after a failed hot-swap, **evaluations are silently served by a provider whose `initialize()` threw**, while `providerMetadata`, `providerRef`, and the status machine all claim the old provider is active. The rollback contract ("a failed swap reverts to the previous working provider") is half-implemented. Severity is bounded (only hit on a failed swap), but `fromAcquireAsync` makes failed swaps a *designed-for* path — its whole point is falling back when the real provider is sick — so the broken half is load-bearing.

## Root-cause anatomy (verified against SDK 1.21.0 sources)

All references are `dev.openfeature:sdk:1.21.0`.

1. **The SDK binds before init and never reverts.** `ProviderRepository.prepareAndInitializeProvider` puts the new provider's `FeatureProviderStateManager` into the domain/default slot *inside* the synchronized block, before initialization ("provider is set immediately, on this thread"). If `initialize()` then throws, `afterError` (`emitErrorAndThrow`) emits `PROVIDER_ERROR` and rethrows — **the slot binding is not reverted**. The SDK keeps routing `client.get*Details` to the failed provider.

2. **The old provider is left alive, unbound, and still attached.** `shutDownOld(oldManager)` sits after `initialize()` in the `try`, so on the failed path it never runs: the old provider is not shut down, not deregistered from the SDK's global provider registry, and — if it extends `EventProvider` — its `attach` state (a CAS on an `AtomicReference<Attachment>`) is never reset via `detach()`.

3. **Why naïve rollback throws.** Re-registering the old provider (`api.setProviderAndWait(old)`) finds no existing state manager for it (its manager was evicted from the slot in step 1), so the SDK calls `attachEventProvider(old)` → `EventProvider.attach(...)` → the CAS fails → `IllegalStateException: Provider X is already attached.` The `detach()` that would clear this is **package-private** in `dev.openfeature.sdk.EventProvider`.

4. **A load-bearing nuance for the fix:** the "already attached" exception fires *exactly and only* in the attached-but-unbound state a failed swap creates. If the old provider is still bound to another slot of the same API (registry topologies can bind one provider instance to several domains), `getExistingStateManagerForProvider` finds it, `attach` is never called, its (READY) state manager is reused, and **re-init is skipped**. So "catch the ISE, detach, retry once" is topology-safe without any knowledge of the SDK's internal domain map — and never severs event propagation for a provider that is still legitimately bound elsewhere.

## Solution design

Two parts: a compile-time package shim exposing `detach`, and a rollback re-registration step in `setProvider`'s failure path.

### Part 1: package shim `EventProviderAccess` (new file)

`core/src/main/scala/dev/openfeature/sdk/EventProviderAccess.scala` — same precedent as the existing `OpenFeatureAPIFactory.scala` sitting right next to it (which already accesses the package-private `OpenFeatureAPI` constructor and works under the same classloader assumptions), and as `ObservingOptimizelyHttpClient` (#267):

```scala
package dev.openfeature.sdk

/** Access to package-private SDK members needed for hot-swap rollback (#282).
  * Compile-time checked: an SDK upgrade that changes these signatures fails compilation
  * here instead of surfacing as a runtime reflection error.
  */
object EventProviderAccess {

  /** Reset an EventProvider's attach state so it can be re-registered with an API instance.
    * No-op for providers that don't extend EventProvider (they never attach).
    * The SDK's detach() is idempotent (plain AtomicReference.set(null)).
    */
  def detach(provider: FeatureProvider): Unit = provider match {
    case ep: EventProvider => ep.detach()
    case _                 => ()
  }

  /** Release a provider from the SDK's global provider registry (spec 1.8.4) — used only in the
    * rollback-failure cleanup path, mirroring what ProviderRepository.shutDownOld would have done.
    */
  def deregisterGlobalProvider(api: OpenFeatureAPI, provider: FeatureProvider): Unit =
    api.deregisterGlobalProvider(provider)
}
```

Do **not** use runtime reflection (`setAccessible`): it breaks silently on SDK upgrades and under JPMS tightening, whereas the shim breaks loudly at compile time. Both approaches were prototyped and pass the existing hot-swap suite; the shim is the robust one.

### Part 2: rollback re-registration in `FeatureFlagsLive.setProvider`

Extend the existing `tapError` (which already rolls back `providerRef`/`providerNameRef` and applies `Signal.SwapFailed`) with an SDK-reconciliation step. Sketch:

```scala
.tapError { _ =>
  providerRef.set(oldProvider) *>
    ZIO.succeed(providerNameRef.set(oldName)) *>
    applySignal(Signal.SwapFailed) *>
    restoreSdkRegistration(oldProvider, newProvider).foldZIO(
      e =>
        // Rollback itself failed: log loudly, best-effort cleanup, leave status Error.
        ZIO.logErrorCause(
          s"setProvider rollback: failed to re-register previous provider '$oldName'; " +
            "status remains Error",
          Cause.fail(e)
        ) *> cleanupUnboundFailedProvider(newProvider),
      restored => applySignal(Signal.RollbackSucceeded).when(restored).unit
    )
}
```

with:

```scala
/** Re-register `oldProvider` with the SDK iff the failed swap left the SDK routing to
  * `newProvider`. Returns true iff the SDK is verified to route to `oldProvider` afterwards.
  */
private def restoreSdkRegistration(
  oldProvider: OFFeatureProvider,
  newProvider: OFFeatureProvider
): Task[Boolean] =
  ZIO.attemptBlocking {
    val bound = domain.fold(api.getProvider())(api.getProvider)
    if (bound eq oldProvider) true        // swap failed before binding (e.g. attach threw); nothing to do
    else if (bound ne newProvider) false  // a concurrent sibling swap won the slot; leave it alone
    else {
      def register(): Unit = domain match {
        case Some(d) => api.setProviderAndWait(d, oldProvider)
        case None    => api.setProviderAndWait(oldProvider)
      }
      try register()
      catch {
        case _: IllegalStateException =>
          // Exactly the attached-but-unbound state a failed swap leaves an EventProvider in
          // (see anatomy §4). Detach and retry once; any further failure propagates.
          EventProviderAccess.detach(oldProvider)
          register()
      }
      true
    }
  }
```

Design decisions, in order of importance:

1. **Guard on the actually-bound provider (`bound eq newProvider`)** before touching the SDK. This makes rollback a no-op when the forward failure happened *before* the slot was rebound (e.g. `registerGlobalProvider`/`attach` threw for the new provider — old provider still bound and serving), and stays out of the way if a concurrent registry-sibling swap has already claimed the slot. When `bound eq oldProvider`, report success — the instance is exactly as before the swap attempt.

2. **Catch-detach-retry rather than detach-upfront.** Per anatomy §4, unconditional detach would sever event propagation for an old provider still bound to another domain of a shared API (registry case) — and the SDK would *not* re-attach it, because the existing-manager path skips `afterSet`. The ISE-triggered detach is precise: it only fires in the pathological state that needs it. (`registerGlobalProvider`'s "bound to another API instance" ISE and the repository's "shutting down" ISE both rethrow identically on retry and propagate — the extra `detach` in those paths is harmless because the provider is being torn down or was never ours.)

3. **Status: `Ready` after a verified rollback.** Rename `Signal.ForceReady` → `Signal.RollbackSucceeded` (same transition, `Some(Ready)`, lifecycle-authoritative) and apply it from `setProvider` itself. PR #281 already established this exact semantic ("restore Ready for the usable fallback") — this moves it from a `fromAcquireAsync`-local patch to the one place that knows the rollback outcome. `setProvider` **still fails** with `ProviderInitializationFailed(originalError)` regardless — the caller sees the failure; status describes the *currently serving* provider, which `setProviderAndWait(old)` has just re-verified. Observers of `statusRef.changes` see a truthful `Error → Ready` recovery; the event hub sees the SDK-emitted `PROVIDER_ERROR` (failed new) followed by `PROVIDER_READY` (restored old). On rollback failure, status remains `Error` (from `SwapFailed`) — also truthful.

4. **Then delete the #281 workaround**: remove `FeatureFlagsLive.forceReady` and the `.tapError(_ => ff.forceReady)` in `FeatureFlags.fromAcquireAsync` (~line 1181); rollback now restores both routing *and* status. Update the stale comment referencing `forceReady` in `NonBlockingInitSpec` (~line 147).

5. **Event-bridge interplay needs no changes.** The rollback runs inside the swap (gate `swapInProgress` still up), so the SDK-emitted `PROVIDER_ERROR` (stamped with the new provider's name) and the rollback's `PROVIDER_READY` (old name, passes the identity guard since refs are already rolled back) are published to the hub but cannot drive the status machine — the rollback's outcome is set explicitly via `RollbackSucceeded`, exactly like the forward path's `SwapSucceeded`.

6. **Failed-provider cleanup.** On *successful* rollback the SDK does it for us: re-registering old triggers `shutDownOld(failedNewManager)` → deregisters the failed provider from the global registry and shuts it down (async on the SDK's task executor). Only when rollback fails during `old.initialize` is the failed new provider left unbound and never shut down — `cleanupUnboundFailedProvider` covers that: if `domain.fold(api.getProvider())(api.getProvider) ne newProvider`, best-effort `ZIO.attemptBlocking(newProvider.shutdown()).ignore` plus `EventProviderAccess.deregisterGlobalProvider(api, newProvider)` (so the instance could later be registered with a different API). If the failed provider is somehow still bound, do nothing.

### Accepted caveats (document in scaladoc + CHANGELOG)

- **Rollback re-initializes the old provider** (fresh `FeatureProviderStateManager` starts `NOT_READY`, so the SDK calls `initialize()` again). An Optimizely poller restarts, an OFREP client re-fetches, etc. Inherent to any SDK-registration-based recovery; the exception is an old provider still bound to another domain, whose READY manager is reused and skips re-init.
- **If the old provider's re-`initialize()` throws, rollback fails** — but note the SDK slot was already rebound to the old provider *before* its init ran, so even this path usually restores routing consistency (`providerRef` == SDK slot == old, both erroring honestly). Status stays `Error`; evaluations proceed per spec 1.7.6 and fail/serve on the old provider's own terms. Only an ISE-retry failure (double-rare) leaves the original mismatch, now loudly logged instead of silent.
- **A tiny TOCTOU window** exists between the `bound eq newProvider` check and re-registration when registry siblings swap the same domain concurrently; each SDK call is serialized under the API's write lock, and the guard makes us stand down rather than fight, so the worst case is a skipped rollback with a logged status.
- **Evaluations concurrent with the rollback window** may still hit the failed provider until `setProviderAndWait(old)` rebinds the slot — bounded by the rollback duration; unavoidable without pausing evaluation.

### Rejected alternatives

- **Runtime reflection on `detach()`** — works today, breaks silently on SDK upgrade / JPMS tightening. The shim gives the same access compile-checked.
- **Permanent routing-facade provider** (register one delegating `EventProvider` wrapper once; swaps become internal ref updates + manual init/shutdown) — makes routing disagreement structurally impossible, but reimplements the SDK's state-manager/lifecycle/event semantics, breaks the #244 provider-identity machinery (events would all carry the wrapper's name), and has a large blast radius across registry/shared-API topologies. Not worth it for a rare failure path.
- **Routing evaluations off `providerRef` directly** (bypass the SDK client on rollback) — loses provider hooks, SDK state checks, and event stamping; diverges two evaluation paths permanently to fix a transient state.

## Test plan

Update (behavior change, assert the *new* truth):

- `ProviderHotSwapSpec` "failed swap sets status to Error" → rename to "failed swap rolls back routing and status to the previous provider": `setProvider` fails with `ProviderInitializationFailed`, **evaluation returns provider A's value** (the regression assertion this issue asks for), `providerStatus == Ready`, `providerMetadata.name == "A"`.
- `ProviderHotSwapSpec` "recovery after failed swap": mid-test status assertion `Error` → `Ready`; rest unchanged (swap to C afterwards must still work — exercises re-registration after rollback).
- `NonBlockingInitSpec` AC2b: strengthen `v.isRight` to `v == Right(false)` — the *fallback's* value, not the failed provider's `true`. This is the original evidence case; it must now pass on values, not just evaluability.

New:

- **Rollback failure**: old provider whose `initialize()` succeeds first, throws on re-init → `setProvider` fails with the *original* error, status `Error`, no hang; evaluation still routes to the old provider (slot was rebound before its init ran).
- **Re-init caveat pinned**: count old provider `initialize()` calls == 2 after a failed swap (documents the poller-restart caveat as intended behavior).
- **EventProvider re-attach**: old provider extending `EventProvider`; after failed swap + rollback, `emitProviderConfigurationChanged` still reaches `ff.events` subscribers (proves detach→re-attach round-trip didn't sever propagation).
- **Failed provider shutdown**: failing provider's `shutdown()` eventually invoked after rollback (SDK `shutDownOld`, async — assert eventually).
- **Shared-provider topology (registry)**: one provider instance bound to two domains via `FeatureFlagRegistry`; failed swap on domain 1 + rollback must not detach the shared provider (its events on domain 2 keep flowing) and must not re-initialize it (READY manager reused). Guards decision 2.
- **Domain and no-domain variants** of the core rollback test (the swap code branches on `domain`).

Existing hot-swap suites (`ProviderHotSwapSpec`, `ProviderHotSwapStressSpec`, `NonBlockingInitSpec`, `FeatureFlagsShutdownSpec`, registry lifecycle #276 test) must stay green; both prototypes of this fix already passed all 12 hot-swap tests.

## Implementation checklist

1. Add `core/src/main/scala/dev/openfeature/sdk/EventProviderAccess.scala` (shared main `scala/` dir — version-agnostic, like `OpenFeatureAPIFactory`).
2. `ProviderStatusMachine`: rename `Signal.ForceReady` → `Signal.RollbackSucceeded`, update scaladoc.
3. `FeatureFlagsLive`: add `restoreSdkRegistration` + `cleanupUnboundFailedProvider`, wire into `setProvider`'s `tapError`; delete `forceReady`.
4. `FeatureFlags.fromAcquireAsync`: drop `.tapError(_ => ff.forceReady)` and its comment.
5. Tests per plan above; update the `NonBlockingInitSpec` comment referencing `forceReady`.
6. CHANGELOG: bug-fix entry noting the behavior change (post-failed-swap status is now `Ready` when rollback succeeds; evaluations revert to the previous provider).
7. `sbt scalafmtAll`, then full verification: `sbt compile scalafmtCheckAll test`.
